src/python/pants/backend/python/dependency_inference/module_mapper.py

```diff
@@ -386,24 +416,38 @@ async def map_module_to_address(
     first_party_mapping: FirstPartyPythonModuleMapping,
     third_party_mapping: ThirdPartyPythonModuleMapping,
 ) -> PythonModuleOwners:
-    providers = [
+    possible_providers: tuple[PossibleModuleProvider, ...] = (
         *third_party_mapping.providers_for_module(request.module, resolve=request.resolve),
         *first_party_mapping.providers_for_module(request.module, resolve=request.resolve),
-    ]
-    addresses = tuple(provider.addr for provider in providers)
+    )
 
-    # There's no ambiguity if there are only 0-1 providers.
-    if len(providers) < 2:
-        return PythonModuleOwners(addresses)
+    # We attempt to disambiguate conflicting providers by taking only the ones for the
+    # closest ancestors to the requested modules. This prevents issues with namespace
+    # packages that are split between first-party and third-party
+    # (e.g., https://github.com/pantsbuild/pants/discussions/17286).
+
+    # Map from provider type to mutable pair of
+    # [closest ancestry found, list of provider of that type at that ancestry level].
+    type_to_closest_providers: dict[ModuleProviderType, list] = defaultdict(lambda: [999, []])
+    for possible_provider in possible_providers:
+        val = type_to_closest_providers[possible_provider.provider.typ]
+        if possible_provider.ancestry < val[0]:
+            val[0] = possible_provider.ancestry
+            val[1] = []
+        # NB This must come after the < check above, so we handle the possible_provider
+        # that caused that check to pass.
+        if possible_provider.ancestry == val[0]:
+            val[1].append(possible_provider.provider)
+
+    closest_providers: list[ModuleProvider] = list(
+        itertools.chain(*[val[1] for val in type_to_closest_providers.values()])
+    )
+    addresses = tuple(provider.addr for provider in closest_providers)
 
-    # Else, it's ambiguous unless there are exactly two providers and one is a type stub and the
-    # other an implementation.
-    if len(providers) == 2 and (providers[0].typ == ModuleProviderType.TYPE_STUB) ^ (
-        providers[1].typ == ModuleProviderType.TYPE_STUB
-    ):
-        return PythonModuleOwners(addresses)
```

Comment:
I'm not following what happens with `ModuleProviderType` now?

Reply:
Line 447 checks that we have at most one value for each type at the closest found ancestry for that type. I'll add a comment to clarify.